### ISSUE TITLE
The Offset Infinite Loop

### ISSUE THREAD

**Issue by vhdmsm** [yagop/node-telegram-bot-api#36]
Hi
My bot has been stuck in infinite loop and it's not responding anymore, what's problem?
here is the stack trace:

```
Unhandled rejection Error: 403 {"ok":false,"error_code":403,"description":"Error: Bot was kicked from a chat"}
    at /Users/vahid/Desktop/node_modules/node-telegram-bot-api/src/telegram.js:72:15
    at tryCatcher (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:503:31)
    at Promise._settlePromiseAt (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:577:18)
    at Promise._settlePromises (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:693:14)
    at Async._drainQueue (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:367:17)
Unhandled rejection Error: 403 {"ok":false,"error_code":403,"description":"Error: Bot was kicked from a chat"}
    at /Users/vahid/Desktop/node_modules/node-telegram-bot-api/src/telegram.js:72:15
    at tryCatcher (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:503:31)
    at Promise._settlePromiseAt (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:577:18)
    at Promise._settlePromises (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:693:14)
    at Async._drainQueue (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/vahid/Desktop/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:367:17)
```

**Comment by mvalipour** [comment]
Any luck fixing this?

**Comment by LostSenSS** [comment]
I have the same problem

**Comment by vhdmsm** [comment]
Actually i've changed my method from polling to telegram webhooks, and my problems are gone now.

**Comment by witnessmenow** [comment]
I have the same issue as well:

My bot is constantly just getting the same message back:

Unhandled rejection Error: 403 {"ok":false,"error_code":403,"description":"[Error]: Forbidden: bot was kicked from the group chat"}
    at /Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/src/telegram.js:121:15
    at tryCatcher (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:507:31)
    at Promise._settlePromiseAt (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:581:18)
    at Promise._settlePromises (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:697:14)
    at Async._drainQueue (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:368:17)
Unhandled rejection Error: 403 {"ok":false,"error_code":403,"description":"[Error]: Forbidden: bot was kicked from the group chat"}
    at /Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/src/telegram.js:121:15
    at tryCatcher (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:507:31)
    at Promise._settlePromiseAt (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:581:18)
    at Promise._settlePromises (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/promise.js:697:14)
    at Async._drainQueue (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/brian/my-reaction-when-telegram-bot/node_modules/node-telegram-bot-api/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:368:17)

**Comment by yagop** [comment]
Seems you are trying to send a message to non exists peer or haven't talk
to you before

**Comment by witnessmenow** [comment]
Sorry your right the stack is due to that, but the reason the issue is appearing is that the bot is constantly receiving the same message over and over again, even thought this user only sent "/next" once, when I start up my bot all I receive is that message, same message_id everytime as well

Does that make more sense?

**Comment by yagop** [comment]
The error causes it won't be marked as read

**Comment by mvalipour** [comment]
@witnessmenow IMPOV, using polling is very unreliable and not worth it. You can very easily switch to webhooks. -- see my blog post: http://mvalipour.github.io/node.js/2015/12/06/telegram-bot-webhook-existing-express/

**Comment by witnessmenow** [comment]
@yagop Is there anyway to move past this error when it occurs? 

Apologies if there is something straight forward I'm missing, I'm pretty new to node and even newer to bot development. 

@mvalipour thanks for the link, looks very straight forward. I had planned on giving the bot a web interface at some stage anyways so would have needed to add an express part down the line as it was. The bot is still in development though, so it wont solve my current problem

**Comment by witnessmenow** [comment]
My specific issue has to be a telegram one, Even if I request a higher offset message directly to the Telegram API its not clearing the messages.

I tweeted telegram hopefully they get back to me.

https://twitter.com/witnessmenow/status/686341804799733760

**Comment by idchlife** [comment]
Same error guys. Bot receives same message with same message_id and does not change this.offset here

https://github.com/yagop/node-telegram-bot-api/blob/master/src/telegramPolling.js#L32

Loop is starting over and over again, I was testing it with breakpoints in WebStorm and it behaves strangely. Always index = 0 and loop restarts again.

I think we should try to find the problem and fix it guys.

@yagop in which case message wont be marked as read?

**Comment by witnessmenow** [comment]
I think its a telegram error, as I was even trying to clear the message using the telegram api directly and couldn't.

It will clear its self in a couple of days

**Comment by idchlife** [comment]
@witnessmenow I just used telegram, manually entered offset like this:

    /getUpdates?offset=415204539

+1-ing every time I got update_id in json. So after some iterations I got rid of old messages.

Problem is not with telegram I think. Something wrong with incrementing update_id in code. If I encounter this problem again I will try to find a solution and create PR.

**Comment by witnessmenow** [comment]
If you look at my tweet from earlier I tried that and it didn't work. Messages for bots clear after 24 hours, did you just happen to wait that long maybe?

**Comment by idchlife** [comment]
@witnessmenow which update_id did you use? I got one from their json and just incremented it and entered and repeated until no old messages left. Maybe they are clearing them automatically but you can erase them manually if you have stuck messages?

Also you may be right. I have this problem since yesterday (don't really remember time). I'll try to create another one.

**Comment by idchlife** [comment]
Yep. I did create another 'situation'.

Recipe is: turn off bot. Send several same messages for which you created callback onText and you get this:

    Got message. Id: 1476 and time: 1454681507065
    Got message. Id: 1476 and time: 1454681507540
    Got message. Id: 1476 and time: 1454681508017
    Got message. Id: 1476 and time: 1454681508480
    Got message. Id: 1476 and time: 1454681508938
    Got message. Id: 1476 and time: 1454681509399

After this I manually entered url which is used inside this bot 

    https://api.telegram.org/botNUMBER:TOKEN/getUpdates?offset=415204539

(offset is kinda random)

I got update_id like this:

    {"ok":true,"result":[{"update_id":415204547,

After it I incremented this update_id and sent it via offset parameter

One by one stacked messages started to disappear. Until I got empty array and bot resumed to work properly. Exactly what this bot is supposed to do when reading updates. This is kinda major bug I think, because you can that easily create it.

We need to go deeper and investigate the code :smiley:

**Comment by lucaslg26** [comment]
+1
Same problem here, I need flush my bot manually incrementing +1 on offset manually.

**Comment by yagop** [comment]
This should be fixed by commit 178f9cd9051b63377fcbb3a13f564b175b1cd3e7. The matter was it wasn't updating the offset if there was an error and can't reach the last message. Now every message processed updates the offset.

**Comment by idchlife** [comment]
Thank you @yagop!

**Comment by lucaslg26** [comment]
Working perfectly now! Thanks @yagop!

**Comment by phillipadsmith** [comment]
@yagop  :100:

**Comment by omnidan** [comment]
@yagop can you please release a new version with this fix? I think the issue here: https://github.com/yagop/node-telegram-bot-api/issues/15#issuecomment-214911783 is related to this, but I noticed 0.20.1 doesn't have this fix yet and there hasn't been a release since then.

**Comment by phillipadsmith** [comment]
> @yagop can you please release a new version with this fix? I think the issue here: #15 (comment) is related to this, but I noticed 0.20.1 doesn't have this fix yet and there hasn't been a release since then.

@omnidan Upgrading to 0.20.1 did fix the issue for me, I believe.

**Comment by omnidan** [comment]
@phillipadsmith the fix isn't part of 0.20.1 (it's from Feb 1 and the fix was made on Mar 13) - I also contacted @yagop a while ago and he told me he meant to release a new version but forgot. Still no new release now 😞

**Comment by phillipadsmith** [comment]
@omnidan Bah, you're totally right.

And, having managed to get into this infinite loop again, I'm wondering if @yagop's commit https://github.com/yagop/node-telegram-bot-api/commit/178f9cd9051b63377fcbb3a13f564b175b1cd3e7 actually fixed it? 

I've just tried using that version of telegramPolling.js and -- if there's a failure -- the incrementing doesn't appear to happen, and the loop begins as far as I can tell. 

Have others resolved this with a patch?  cc @idchlife @lucaslg26

**Comment by RacerBas** [comment]
I'm having the same issue right now @phillipadsmith

**Comment by GochoMugo** [comment]
This is what I have gathered so far:

---

#### when and how does the bug occur?

The bug occurs when an error is thrown during processing of an update.

During this processing of an update (at [src/telegramPolling.js#L46](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegramPolling.js#L46)), an error could be thrown while handling a [regexp callback](https://github.com/yagop/node-telegram-bot-api#TelegramBot+onText) (at [src/telegram.js#L112](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegram.js#L112)) or when handling a [onReplyToMessage callback](https://github.com/yagop/node-telegram-bot-api#telegrambotonreplytomessagechatid-messageid-callback) (at [src/telegram.js#L126](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegram.js#L126)). Such an error would bubble up and be handled by the enveloping promise (see [src/telegramPolling.js#L40](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegramPolling.js#L40)). The promise would simply re-throw the error (at [src/telegramPolling.js#L51](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegramPolling.js#L51)). This results in the "Unhandled rejection Error" as [seen above](https://github.com/yagop/node-telegram-bot-api/issues/36#issue-109630670). So far, such **unhandled** errors do **NOT** terminate the Node.js process. The unterminated process will continue polling (see [src/telegramPolling.js#L58](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegramPolling.js#L58))!

However, @yagop's [fix](https://github.com/yagop/node-telegram-bot-api/commit/178f9cd9051b63377fcbb3a13f564b175b1cd3e7), updates the offset **before** processing an update. Since [an update is considered confirmed as soon as getUpdates is called with an offset higher than its update_id](https://core.telegram.org/bots/api#getupdates), **all** the updates received, including the update that caused the chain of errors to occur, would be confirmed, as soon as we enter the next poll interval (see [src/telegramPolling.js#L94](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegramPolling.js#L94)). Thus, the fix (kinda) works!

**Before the fix**, the offset would **not** have been updated (see old code at [src/telegramPolling.js#L34](https://github.com/yagop/node-telegram-bot-api/blob/610a989e69b42cc86c3a3c6bf7643ac7529fb61a/src/telegramPolling.js#L34)), causing the next poll interval to retrieve already-processed updates, including the update that originally caused the error! Thus **LOOP!**

Note that **if the process terminated** and we restarted it again, this bug would re-surface since we would **not** have marked the already-processed updates as confirmed! This is **important** considering that in the future, unhandled promise rejections will terminate the process. Thus, the fix will stop working some time in the future!

In some cases, the bug would disappear, as the offending update would have been [discarded by Telegram after their 24hr limit](https://core.telegram.org/bots/api#getting-updates)!

---

#### fixing:

It is currently fixed!!!

The current fix works, as long as unhandled promise rejections do **not** terminate the process. Once Node.js behaves by terminating, this loop will re-surface! To prevent this Loop completely, we need a way to consider this:

1. Processed updates **MUST** be marked confirmed
1. The update causing the error **may** remain unconfirmed to allow re-processing (we need more discussion around this)

To achieve the above, in our `.catch()` (at [src/telegramPolling.js#L49](https://github.com/yagop/node-telegram-bot-api/blob/2a4fea42d71d9e6b221414ed25e5cc8b601279cf/src/telegramPolling.js#L49)), we may do the following, in order:

1. send a `getUpdates` request with the updated `offset + 1`, therefore marking all processed updates as confirmed!
1. re-throw the error, thus terminating the process!

---

#### notes:

**Reproducing this bug** is important. If you believe that this bug has **not** been fixed, please provide an **elaborate** way to reproduce it. (@yagop [had a hard time doing so](https://github.com/yagop/node-telegram-bot-api/issues/15#issuecomment-125043346))

---

#### related issues:

* Offset Issue if Error: #15

**Comment by GochoMugo** [comment]
> The current fix works, as long as unhandled promise rejections do not terminate the process. Once Node.js behaves by terminating, this loop will re-surface! To prevent this Loop completely...
>
> &mdash; [comment above](https://github.com/yagop/node-telegram-bot-api/issues/36#issuecomment-268532067)

PR #265 handles (future) cases where unhandled promise rejections will terminate the process. Please proceed to the PR to leave comments on the implementation!

**Comment by kamikazechaser** [comment]
👍For fixing this!

**Comment by luckydonald** [comment]
@mvalipour link is down

**Comment by mvalipour** [comment]
@luckydonald Fixed it. damn trailing slash :/

**Comment by GochoMugo** [comment]
Fixed in **v0.30.0**.
